src/base/LruMap.h

```diff
@@ -148,39 +169,47 @@ LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
     if (memLimit() == 0)
         return false;
 
-    index.push_front(new Entry(key, t));
+    auto e = new Entry(key, t, ttl);
+    memUsed_ += e->memCost() + e->key.length() + sizeof(Entry) + sizeof(MapPair);
+    index.push_front(e);
     storage.insert(MapPair(key, index.begin()));
 
     ++entries_;
+
+    // if over memory limit, start purging
+    while (memUsed_ > memLimit_)
+        del(storage.end());
+
     return true;
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t EntryCost(const EntryValue *)>
 bool
 LruMap<EntryValue, EntryCost>::expired(const LruMap::Entry &entry) const
 {
-    if (ttl < 0)
+    if (retention < 0)
         return false;
 
-    return (entry.date + ttl < squid_curtime);
+    return (entry.expires < squid_curtime);
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t EntryCost(const EntryValue *)>
 bool
 LruMap<EntryValue, EntryCost>::del(LruMap::MapIterator const &i)
 {
     if (i != storage.end()) {
         Entry *e = *i->second;
         index.erase(i->second);
         storage.erase(i);
+        memUsed_ -= (e->memCost() + e->key.length() + sizeof(Entry) + sizeof(MapPair));
```

Comment:
Adding an Entry method to report memory usage is better than adding a `memoryUsedBy(const Entry&)` method that I suggested -- I forgot that Entry is not EntryValue; we do control the Entry class and should add methods to it, of course.

I think we should standardize the memory usage terminology to avoid a mixture of "cost"- and "use(d)"-based terms. I do not have a strong preference: "cost" works slightly better for predicted/future use while "use" works better for current use. Please try to eliminate the mixture one way or the other as you polish/add code.